### PR TITLE
[WIP] baremetal: Pass through of config variables to ironic

### PR DESCRIPTION
**Pull request by derekhiggins**
Makes use of oslo.config's ability to read config from env variables
see https://github.com/metal3-io/ironic-image/pull/165

### CODE REVIEW

**Comment by derekhiggins** [comment]
/hold
this only works for ironic on the bootstrap node, holding while I look into how to also do it on master nodes.

File: pkg/types/baremetal/platform.go [hunk 1]
```diff
@@ -108,4 +108,7 @@ type Platform struct {
 	//
 	// +optional
 	ClusterOSImage string `json:"clusterOSImage,omitempty" validate:"omitempty,osimageuri,urlexist"`
+
+
+	IronicExtraConf map[string]interface{} `json:"ironicExtraConf,omitempty"`
```
Comment (stbenjam) [inline pkg/types/baremetal/platform.go:113]: Public struct members need godoc comment

**Comment by stbenjam** [comment]
>  this only works for ironic on the bootstrap node, holding while I look into how to also do it on master nodes.

We'll need to update the Provisioning CRD to support a mechanism to pass-through Ironic configuration. It'll be an API change.  https://github.com/openshift/machine-api-operator/pull/632 alters the API and makes use of it, not quite the same idea but it should give you the gist of how the change would look.

File: pkg/types/baremetal/platform.go (hunk 1 above)
Comment (abhinavdahiya) [inline pkg/types/baremetal/platform.go:113]: secondly a map of string to _anything_ is not a great API.

Let's make sure there is an enhancement for this new field before we move ahead on this.

/hold

Comment (derekhiggins) [reply pkg/types/baremetal/platform.go:113]: > secondly a map of string to _anything_ is not a great API.
> 
> Let's make sure there is an enhancement for this new field before we move ahead on this.
> 
> /hold

Agreed, mapping to _anything_ is not ideal, the problem is that ironic has ~700 config options and ironic-inspector has another ~400, in order to allow people to deal with special cases we either need to 
1. have some sort of pass through mechanism like this or 
2. add individual options for each one as it comes up with the delay thats involved to merge/release

as is we're seeing patches merged into the ironic image with options specific to CI, specific hardware and for debugging
https://github.com/metal3-io/ironic-image/pull/172
https://github.com/metal3-io/ironic-image/pull/160
https://github.com/metal3-io/ironic-image/pull/167
https://github.com/metal3-io/ironic-image/pull/150

some of these could be top-level platform configuration options but for corner cases I don't think we'd want to start creating options for every single one.

I can create an enhancement for this, I didn't intend on removing the hold on this PR until we came to a consensuses anyways.

**Comment by derekhiggins** [comment]
> I'm not sure I like the idea of a top-level platform configuration for setting Ironic's internal configuration. It's exposing an implementation detail we prefer to keep hidden, if at all possible. If we have to make it available, it might belong tucked away in an `Advanced` struct member in the baremetal platform, with appropriate warnings.
Ack, will look into it

> 
> This also doesn't handle how we pass this through to MAO/CBO for the in-cluster Ironic images.
yup, I was going to look into that next

> 
> TBH, I think this change needs an enhancement request so we can get all the details worked out.
ack, will also look into creating one



> I'm not sure I like the idea of a top-level platform configuration for setting Ironic's internal configuration. It's exposing an implementation detail we prefer to keep hidden, if at all possible. If we have to make it available, it might belong tucked away in an `Advanced` struct member in the baremetal platform, with appropriate warnings.
> 
> This also doesn't handle how we pass this through to MAO/CBO for the in-cluster Ironic images.
> 
> TBH, I think this change needs an enhancement request so we can get all the details worked out.

**Comment by derekhiggins** [comment]
> > this only works for ironic on the bootstrap node, holding while I look into how to also do it on master nodes.
> 
> We'll need to update the Provisioning CRD to support a mechanism to pass-through Ironic configuration. It'll be an API change. [openshift/machine-api-operator#632](https://github.com/openshift/machine-api-operator/pull/632) alters the API and makes use of it, not quite the same idea but it should give you the gist of how the change would look.

thanks, I was trying something similar, this should help me get it right.

**Comment by derekhiggins** [comment]
@stbenjam @abhinavdahiya I've started an enhancement here if ye have any thought I can update it,
https://github.com/openshift/enhancements/pull/401

**Comment by derekhiggins** [comment]
OK will close this, looks like we'd rather add things when needed.

**Pull request closed without merging**